alpenhorn/io/_default_asyncs.py

```diff
@@ -2,19 +2,168 @@
+
+    # Giddy up!
+    start_time = time.time()
+    log.info(f'Transferring file "{req.file.path}":')
+
+    # Attempt to transfer the file. Each of the methods below needs to return
+    # a dict with required key:
+    #  - ret : integer
+    #        return code (0 == success)
+    # optional keys:
+    #  - md5sum : string or True
+    #        If True, the sum is guaranteed to be right; otherwise, it's a
+    #        md5sum to check against the source.  Must be present if ret == 0
+    #  - stderr : string
+    #        if given, printed to the log when ret != 0
+    #  - check_src : bool
+    #        if given and False, the source file will _not_ be marked suspect
+    #        when ret != 0; otherwise, a failure results in a source check
+
+    # First we need to check if we are copying over the network
+    if not local:
+        if shutil.which("bbcp") is not None:
+            # First try bbcp which is a fast multistream transfer tool. bbcp can
+            # calculate the md5 hash as it goes, so we'll do that to save doing
+            # it at the end.
+            ioresult = ioutil.bbcp(from_path, to_dir, req.file.size_b)
+        elif shutil.which("rsync") is not None:
+            # Next try rsync over ssh.
+            ioresult = ioutil.rsync(from_path, to_dir, req.file.size_b, local)
+        else:
+            # We have no idea how to transfer the file...
+            log.error("No commands available to complete remote pull.")
+            ioresult = {"ret": -1, "check_src": False}
+
+    else:
+        # Okay, great we're just doing a local transfer.
+
+        # First try to just hard link the file. This will only work if we
+        # are on the same filesystem.  If it didn't work, ioresult will be None
+        #
+        # But don't do this if it creates a hardlink between an archive node and
+        # a non-archive node
+        if req.node_from.archive == io.node.archive:
+            ioresult = ioutil.hardlink(from_path, to_dir, req.file.name)
+        else:
+            ioresult = None
+
+        # If we couldn't just link the file, try copying it with rsync.
+        if ioresult is None:
+            if shutil.which("rsync") is not None:
+                ioresult = ioutil.rsync(from_path, to_dir, req.file.size_b, local)
+            else:
+                log.error("No commands available to complete local pull.")
+                ioresult = {"ret": -1, "check_src": False}
+
+    # Delete the placeholder, if we created it
+    placeholder.unlink(missing_ok=True)
+
+    if not ioutil.copy_request_done(
+        req,
+        io,
+        check_src=ioresult.get("check_src", True),
+        md5ok=ioresult.get("md5sum", None),
+        start_time=start_time,
+        stderr=ioresult.get("stderr", None),
+        success=(ioresult["ret"] == 0),
+    ):
+        # Remove file, on error
+        try:
+            if to_file.exists():
+                to_file.unlink()
```

Comment:
```suggestion
            to_file.unlink(missing_ok=True)
```
avoiding the (admittedly unlikely) race condition